roachpb: treat GetRequests as non-directional in BatchRequest.Split

yuzefovich opened the pull request:
This commit updates `BatchRequest.Split` to treat `GetRequest`s as
non-directional so that batches with Gets and ReverseScans are not
split. Without this fix currently when reading from a secondary index
in the reverse direction, we might crash. This was exposed recently by
issueing the GetRequests for indexes without STORING clause, but it is
present on 21.2 for index with STORING clause.

Fixes: #73103.

Release note (bug fix): Previously, CockroachDB could crash when reading
of a secondary index with STORING clause in reverse direction (because
of ORDER BY col DESC). The bug was introduced in 21.2.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73272)
<!-- Reviewable:end -->

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-:-MphK-b82Hiz5iSRErS7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/sqllite, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-MphJdds5BjaTMahb4--:-MphJdds5BjaTMahb4-0:b2nrry0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db2a38b8013dd3c0ad750adc52592234aeb98f98/pkg/sql/logictest/testdata/logic_test/sqllite#L473)):*
> ```
> query II
> SELECT pk, col0 FROM t73103_1 WHERE col0 <= 63 OR col0 IN (27,11,93,41) ORDER BY 2 DESC
> ----
> ```

Why does the test use two tables?


<!-- Sent from Reviewable.io -->

jordanlewis commented:
Hmm - do we need to be this restrictive? What is the reason why we can't intertwine GetRequest and RevScan? Can we lift the restriction instead of removing this optimization?

yuzefovich commented:
> Hmm - do we need to be this restrictive? What is the reason why we can't intertwine GetRequest and RevScan? Can we lift the restriction instead of removing this optimization?

That's a good point that I've also wondered about. My guess is that it should be ok to intertwine GetRequests and ReverseScanRequests, but that would require us to change `BatchRequest.Split` method and that allowing that intertwine might lead to violation of some internal invariants that I don't know anything about. I'm hoping that @nvanbenschoten knows whether that would be a good idea.

nvanbenschoten commented:
I don't think that there's a fundamental reason why `GetRequests` and `ReverseScanRequests` can't be combined in the same batch, even with a limit. We do assert [here](https://github.com/cockroachdb/cockroach/blob/a80cfbee826f70988381a0a85c0fe7aba0115484/pkg/kv/kvclient/kvcoord/dist_sender.go#L645) that `ScanRequests` and `ReverseScanRequests` can't be in the same batch with a limit, because it's not clear what a resume span would mean in that case. But this is not the case with `GetRequests`, which aren't "directional".

So this seems like a synthetic limitation of the `BatchRequest.Split` method. That method is splitting `GetRequests` and `ReverseScanRequests` because `ReverseScanRequests` carries the `isReverse` flag while `GetRequests` does not. We could redefine this compatibility condition if we wanted. For instance, we probably don't need to consider direction unless both operations are ranged (have the `isRange` flag), so the following adjustment to that splitting logic might address this:

<details><summary>Possible fix</summary>
<p>

```go
diff --git a/pkg/roachpb/batch.go b/pkg/roachpb/batch.go
index 5bcf5d3cef..50490de14c 100644
--- a/pkg/roachpb/batch.go
+++ b/pkg/roachpb/batch.go
@@ -551,7 +551,10 @@ func (ba BatchRequest) Split(canSplitET bool) [][]RequestUnion {
                // enforcing are that a batch can't mix non-writes with writes.
                // Checking isRead would cause ConditionalPut and Put to conflict,
                // which is not what we want.
-               const mask = isWrite | isAdmin | isReverse
+               mask := isWrite | isAdmin
+               if (exFlags&isRange) != 0 && (newFlags&isRange) != 0 {
+                       mask |= isReverse
+               }
                return (mask & exFlags) == (mask & newFlags)
        }
        var parts [][]RequestUnion
diff --git a/pkg/roachpb/batch_test.go b/pkg/roachpb/batch_test.go
index 02b7c6d031..c396d00d2a 100644
--- a/pkg/roachpb/batch_test.go
+++ b/pkg/roachpb/batch_test.go
@@ -93,6 +93,7 @@ func TestBatchSplit(t *testing.T) {
                {[]Request{get, get, get, put, put, get, get}, []int{3, 2, 2}, true},
                {[]Request{spl, get, scan, spl, get}, []int{1, 2, 1, 1}, true},
                {[]Request{spl, spl, get, spl}, []int{1, 1, 1, 1}, true},
+               {[]Request{rv, get, rv, get}, []int{4}, true},
                {[]Request{get, scan, get, dr, rv, put, et}, []int{3, 1, 1, 1, 1}, true},
                // Same one again, but this time don't allow EndTxn to be split.
                {[]Request{get, scan, get, dr, rv, put, et}, []int{3, 1, 1, 2}, false},
```

</p>
</details>

----

Also, I don't quite follow why this was caused by https://github.com/cockroachdb/cockroach/pull/72794. Wasn't it still possible to hit this even before that PR? For instance, if whatever index that we are using `STORED` a column, then wouldn't we have always been able to hit this with the query in question? Did that PR just make this easier to hit in logictests because we don't often store columns in indexes?

yuzefovich committed:
roachpb: treat GetRequests as non-directional in BatchRequest.Split

This commit updates `BatchRequest.Split` to treat `GetRequest`s as
non-directional so that batches with Gets and ReverseScans are not
split. Without this fix currently when reading from a secondary index
in the reverse direction, we might crash. This was exposed recently by
issueing the GetRequests for indexes without STORING clause, but it is
present on 21.2 for index with STORING clause.

Release note (bug fix): Previously, CockroachDB could crash when reading
of a secondary index with STORING clause in reverse direction (because
of ORDER BY col DESC). The bug was introduced in 21.2.

yuzefovich reviewed (commented):
Indeed, Nathan, you are right - the bug has been pre-existing, and #72794 only exposed it. I used your proposed fix to relax `Split` for this case. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-:-Mps6zL255mJ44fSpsqA:be850v1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @nvanbenschoten)

___
*[pkg/sql/logictest/testdata/logic_test/sqllite, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-MphJdds5BjaTMahb4--:-Mps79r95YSJJmKaiCG7:bbb44av) ([raw file](https://github.com/cockroachdb/cockroach/blob/db2a38b8013dd3c0ad750adc52592234aeb98f98/pkg/sql/logictest/testdata/logic_test/sqllite#L473)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Why does the test use two tables?
</blockquote></details>

Didn't simplify the repro enough, it should be good now.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks Yahor!

> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-:-MpsCcbq9BLwsnA19WL-:b-3tx2u2)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @yuzefovich)

___
*[pkg/roachpb/batch.go, line 564 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-MpsBbEr46IQnWgsOa2C:-MpsBbEr46IQnWgsOa2D:bweisg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f249aa8e59b7fb6ef329a9ff33c8361091beb27/pkg/roachpb/batch.go#L564)):*
> ```Go
> 		if (exFlags&isRange) != 0 && (newFlags&isRange) != 0 {
> 			// Require the same value for isReverse flag only for ranged
> 			// requests.
> ```

Let's say a bit more. Something about why the direction of requests in a batch needs to be the same (hint: see the use of `scanDir` in `divideAndSendBatchToRanges`) but also how non-ranged operations are not directional, and so we only need to concern ourselves with forward vs. reverse requests if the existing and the new requests are both ranged.
___
*[pkg/roachpb/batch_test.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-MpsBcWh21l4tJoLXnDi:-MpsBcWh21l4tJoLXnDj:b-8xpolf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f249aa8e59b7fb6ef329a9ff33c8361091beb27/pkg/roachpb/batch_test.go#L96)):*
> ```Go
> 		{[]Request{spl, get, scan, spl, get}, []int{1, 2, 1, 1}, true},
> 		{[]Request{spl, spl, get, spl}, []int{1, 1, 1, 1}, true},
> 		{[]Request{rv, get, rv, get}, []int{4}, true},
> ```

Could you also add `{[]Request{scam, get, scan, get}, []int{4}, true},` above this test case.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the careful review!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73272#-:-MpsU3XN0e0G1_DywSev:b-ae1x1u)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach and @nvanbenschoten)



<!-- Sent from Reviewable.io -->
